### PR TITLE
KAFKA-13587; Implement leader recovery for KIP-704

### PR DESCRIPTION
**Pull request by jsancio**
Implementation of the protocol for starting and stopping leader recovery after an unclean leader election. This includes the management of state in the controllers (legacy and KRaft) and propagating this information to the brokers. This change doesn't implement log recovery after an unclean leader election.

### Protocol Changes
For the topic partition state znode, the new field "leader_recovery_state" was added. If the field is missing the value is assumed to be RECOVERED.

ALTER_PARTITION was renamed from ALTER_ISR. The CurrentIsrVersion field was renamed to PartitionEpoch. The new field LeaderRecoveryState was added.

The new field LeaderRecoverState was added to the LEADER_AND_ISR request. The inter broker protocol version is used to determine which version to send to the brokers.

A new tagged field for LeaderRecoveryState was added to both the PartitionRecord and PartitionChangeRecord.

### Controller
For both the KRaft and legacy controller the LeaderRecoveryState is set to RECOVERING, if the leader was elected out of the ISR, also known as unclean leader election. The controller sets the state back to RECOVERED after receiving an ALTER_PARTITION request with version 0, or with version 1 and with the LeaderRecoveryState set to RECOVERED.

Both controllers preserve the leader recovery state even if the unclean leader goes offline and comes back online before an RECOVERED ALTER_PARTITION is sent.

The controllers reply with INVALID_REQUEST if the ALTER_PARTITION either:
1. Attempts to increase the ISR while the partition is still RECOVERING
2. Attempts to change the leader recovery state to RECOVERING from a RECOVERED state.

### Topic Partition Leader
The topic partition leader doesn't implement any log recovery in this change. The topic partition leader immediately marks the partition as RECOVERED and sends that state in the next ALTER_PARTITION request.

### Topic Partition Follower
The topic partition follower will reject FETCH requests with NOT_LEADER_OR_FOLLOWER while the leader is RECOVERING. The partition follower can receive FETCH requests because of KIP-392. The partition follower will start accepting requests after the leader has recovered and the follower has received a RECOVERED leader state through the LEADER_AND_ISR or through the partition records in the KRaft mode.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: checkstyle/suppressions.xml [hunk 1]
```diff
@@ -276,6 +276,8 @@
               files="(QuorumController|QuorumControllerTest|ReplicationControlManager|ReplicationControlManagerTest).java"/>
     <suppress checks="ClassFanOutComplexity"
               files="(QuorumController|ReplicationControlManager|ReplicationControlManagerTest).java"/>
+    <suppress checks="MethodLength"
```
Comment (hachikuji) [inline checkstyle/suppressions.xml]: Probably a sign that `alterPartition` is getting out of control. Could we consider adding some helpers instead?

File: clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java [hunk 2]
```diff
@@ -110,6 +110,14 @@
     LIST_TRANSACTIONS(ApiMessageType.LIST_TRANSACTIONS),
     ALLOCATE_PRODUCER_IDS(ApiMessageType.ALLOCATE_PRODUCER_IDS, true, true);
 
+    /**
+     * ALTER_ISR was the old name for ALTER_PARTITION.
+     *
+     * @deprecated since 3.2.0. Use {@link #ALTER_PARTITION} instead
+     */
+    @Deprecated
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java]: Is this necessary? I don't believe this is a public API.

File: clients/src/main/resources/common/message/AlterPartitionRequest.json [hunk 3]
```diff
@@ -34,9 +34,11 @@
...
           "about": "The leader epoch of this partition" },
         { "name": "NewIsr", "type": "[]int32", "versions": "0+", "entityType": "brokerId",
-          "about": "The ISR for this partition"},
-        { "name": "CurrentIsrVersion", "type": "int32", "versions": "0+",
-          "about": "The expected version of ISR which is being updated"}
+          "about": "The ISR for this partition" },
+        { "name": "LeaderRecoveryState", "type": "int8", "versions": "1+", "default": "0",
+          "about": "1 if the partition is recovering from an unclean leader election; 0 otherwise." },
+        { "name": "PartitionEpoch", "type": "int32", "versions": "0+",
+          "about": "The expected epoch of the partition which is being updated" }
```
Comment (hachikuji) [inline clients/src/main/resources/common/message/AlterPartitionRequest.json]: It might be useful to mention that this refers to the zk version for zk clusters? Same question for response.

File: clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java [hunk 4]
```diff
@@ -1305,33 +1305,33 @@ private DescribeUserScramCredentialsResponse createDescribeUserScramCredentialsR
         return new DescribeUserScramCredentialsResponse(data);
     }
 
-    private AlterIsrRequest createAlterIsrRequest(short version) {
-        AlterIsrRequestData data = new AlterIsrRequestData()
+    private AlterPartitionRequest createAlterPartitionRequest(short version) {
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:1308]: Shall we do any tests for the new fields?

File: core/src/main/scala/kafka/api/LeaderAndIsr.scala [hunk 5]
```diff
@@ -52,11 +63,12 @@ case class LeaderAndIsr(leader: Int,
     } else if (other == null) {
       false
     } else {
-      leader == other.leader && leaderEpoch == other.leaderEpoch && isr.equals(other.isr)
+      leader == other.leader && leaderEpoch == other.leaderEpoch && isr.equals(other.isr) &&
+      leaderRecoveryState == other.leaderRecoveryState
```
Comment (hachikuji) [inline core/src/main/scala/kafka/api/LeaderAndIsr.scala]: nit: indent this?

File: core/src/main/scala/kafka/cluster/Partition.scala [hunk 6]
```diff
@@ -147,56 +147,65 @@ sealed trait IsrState {
...
+   * Indicates if we have an AlterPartition request inflight.
    */
   def isInflight: Boolean
 }
 
-sealed trait PendingIsrChange extends IsrState {
+sealed trait PendingPartitionChange extends PartitionState {
   def sentLeaderAndIsr: LeaderAndIsr
+
+  override val leaderRecoveryState: LeaderRecoveryState = LeaderRecoveryState.RECOVERED
```
Comment (hachikuji) [inline core/src/main/scala/kafka/cluster/Partition.scala:163]: I can't say I like the idea of getting this implicitly. I do get the point that we would not send `AlterPartition` with anything except `LeaderRecoveryState.RECOVERED` (for now), but I think we would tend to overlook the implication when it is hidden here. Can we at least include it in the `toString` implementations?

File: core/src/main/scala/kafka/cluster/Partition.scala [hunk 7]
```diff
@@ -542,7 +551,8 @@ class Partition(val topicPartition: TopicPartition,
         assignment = partitionState.replicas.asScala.map(_.toInt),
         isr = isr,
         addingReplicas = addingReplicas,
-        removingReplicas = removingReplicas
+        removingReplicas = removingReplicas,
+        LeaderRecoveryState.RECOVERED
```
Comment (hachikuji) [inline core/src/main/scala/kafka/cluster/Partition.scala:563]: Can we at least log a message if the LeaderAndIsr indicates recovery is needed? I think with this implementation, if there is only one replica, the state would remain RECOVERING indefinitely. Do I have that right?

File: core/src/main/scala/kafka/cluster/Partition.scala [hunk 8]
```diff
@@ -1080,6 +1092,9 @@ class Partition(val topicPartition: TopicPartition,
     // decide whether to only fetch from leader
     val localLog = localLogWithEpochOrException(currentLeaderEpoch, fetchOnlyFromLeader)
 
+    // Check that the partition leader is recovering from an unclean leader election.
+    validateLeaderRecoveryState()
```
Comment (hachikuji) [inline core/src/main/scala/kafka/cluster/Partition.scala]: Are we doing this on writes as well?

File: core/src/main/scala/kafka/controller/Election.scala [hunk 9]
```diff
@@ -40,7 +40,14 @@ object Election {
         val newLeaderAndIsrOpt = leaderOpt.map { leader =>
           val newIsr = if (isr.contains(leader)) isr.filter(replica => controllerContext.isReplicaOnline(replica, partition))
           else List(leader)
-          leaderAndIsr.newLeaderAndIsr(leader, newIsr)
+
+          if (!isr.contains(leader)) {
+            // The new leader is not in the old ISR so mark the partition a RECOVERING
+            leaderAndIsr.newRecoveringLeaderAndIsr(leader, newIsr)
+          } else {
+            // Elect a new leader but keep the previous leader recovery state
```
Comment (hachikuji) [inline core/src/main/scala/kafka/controller/Election.scala:49]: Is it possible for the state to be RECOVERING here?

File: core/src/main/scala/kafka/controller/Election.scala [hunk 10]
```diff
@@ -53,17 +60,17 @@ object Election {
    * Elect leaders for new or offline partitions.
    *
    * @param controllerContext Context with the current state of the cluster
-   * @param partitionsWithUncleanLeaderElectionState A sequence of tuples representing the partitions
+   * @param partitionsWithUncleanLeaderLeaderRecoveryState A sequence of tuples representing the partitions
```
Comment (hachikuji) [inline core/src/main/scala/kafka/controller/Election.scala]: nit: the original name seems fine for what it returns. "LeaderLeader"?

File: core/src/main/scala/kafka/controller/KafkaController.scala [hunk 11]
```diff
@@ -2781,8 +2812,9 @@ case object IsrChangeNotification extends ControllerEvent {
   override def preempt(): Unit = {}
 }
 
-case class AlterIsrReceived(brokerId: Int, brokerEpoch: Long, isrsToAlter: Map[TopicPartition, LeaderAndIsr],
-                            callback: AlterIsrCallback) extends ControllerEvent {
+case class AlterPartitionReceived(
+  brokerId: Int, brokerEpoch: Long, partitionssToAlter: Map[TopicPartition, LeaderAndIsr], callback: AlterPartitionCallback
```
Comment (hachikuji) [inline core/src/main/scala/kafka/controller/KafkaController.scala]: nit: extra 's' in "partitionss"

File: core/src/main/scala/kafka/controller/KafkaController.scala [hunk 12]
```diff
@@ -2263,37 +2271,43 @@ class KafkaController(val config: KafkaConfig,
...
                       .setErrorCode(error.code))
                   case Right(leaderAndIsr) => topicResp.partitions.add(
-                    new AlterIsrResponseData.PartitionData()
+                    new AlterPartitionResponseData.PartitionData()
                       .setPartitionIndex(tp.partition)
                       .setLeaderId(leaderAndIsr.leader)
                       .setLeaderEpoch(leaderAndIsr.leaderEpoch)
                       .setIsr(leaderAndIsr.isr.map(Integer.valueOf).asJava)
-                      .setCurrentIsrVersion(leaderAndIsr.zkVersion))
+                      .setLeaderRecoveryState(leaderAndIsr.leaderRecoveryState.value)
```
Comment (hachikuji) [inline core/src/main/scala/kafka/controller/KafkaController.scala]: Do we need to make this field ignorable in the response?

File: core/src/main/scala/kafka/server/AlterIsrManager.scala [hunk 13]
```diff
@@ -250,23 +260,35 @@ class DefaultAlterIsrManager(
...
+                  LeaderAndIsr(
+                    partition.leaderId,
+                    partition.leaderEpoch,
+                    partition.isr.asScala.toList.map(_.toInt),
+                    LeaderRecoveryState.of(partition.leaderRecoveryState),
+                    partition.partitionEpoch
+                  )
+                )
+              } catch {
+                case e: IllegalArgumentException =>
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/AlterIsrManager.scala]: It might be cleaner to let `LeaderRecoveryState.of` return `Optional<LeaderRecoveryState>`

File: core/src/test/scala/kafka/api/LeaderAndIsrTest.scala [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+import org.apache.kafka.metadata.LeaderRecoveryState
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Test
+
+final class LeaderAndIsrTest {
+  @Test
+  def testRecoveringLeaderAndIsr(): Unit = {
+    val leaderAndIsr = LeaderAndIsr(1, List(1, 2))
+    val recoveringLeaderAndIsr = leaderAndIsr.newRecoveringLeaderAndIsr(3, List(3))
+
```
Comment (hachikuji) [inline core/src/test/scala/kafka/api/LeaderAndIsrTest.scala:29]: nit: extra newline

File: core/src/test/scala/kafka/zk/TopicPartitionStateZNodeTest.scala [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+import org.apache.zookeeper.data.Stat
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Test
+import org.mockito.Mockito.mock
+import org.mockito.Mockito.when
+
+final class TopicPartitionStateZNodeTest {
+
+  @Test
+  def testEncodeDecode(): Unit = {
```
Comment (hachikuji) [inline core/src/test/scala/kafka/zk/TopicPartitionStateZNodeTest.scala]: Maybe check the default case as well when leader recovery state is not defined?

File: metadata/src/main/java/org/apache/kafka/metadata/LeaderRecoveryState.java [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+     */
+    RECOVERING((byte) 1);
+
+    /**
+     * A special value used to represent that the LeaderRecoveryState field of a
+     * PartitionChangeRecord didn't change.
+     */
+    private static final byte NO_CHANGE = (byte) -1;
+
+    private static final Map<Byte, LeaderRecoveryState> VALUE_TO_ENUM;
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/metadata/LeaderRecoveryState.java]: nit: the map seems like overkill with two values

File: clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java (hunk 2 above)
Comment (jsancio) [reply clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java]: Okay. I wasn't sure if the symbols in `common/protocol/ApiKeys.java` were public. I am okay keeping it as it is a light way of keeping backward compatibility just in case external code is using this `enum`.

File: core/src/main/scala/kafka/server/AlterIsrManager.scala (hunk 13 above)
Comment (jsancio) [reply core/src/main/scala/kafka/server/AlterIsrManager.scala]: Yeah. I went back and forth of this. There are some code paths were we want to throw an exception as in most cases this means a bug in Kafka. I think this was the only case where we want to handle this exception. I'll take a look again.

File: core/src/main/scala/kafka/controller/KafkaController.scala (hunk 12 above)
Comment (jsancio) [reply core/src/main/scala/kafka/controller/KafkaController.scala]: It maybe a little subtle but I don't think we need to. If the version is `0` then this is guaranteed to be the default value `0` so the serialization will succeed. This is true because we only write these values in the response when the operation success. If the operation fails then we skip writing these values and instead just write the error code.

File: core/src/main/scala/kafka/server/AlterIsrManager.scala (hunk 13 above)
Comment (jsancio) [reply core/src/main/scala/kafka/server/AlterIsrManager.scala]: Okay. We have both: `of` which throws and `optionalOf` which returns an `Optional`.

File: metadata/src/main/java/org/apache/kafka/metadata/LeaderRecoveryState.java (hunk 16 above)
Comment (jsancio) [reply metadata/src/main/java/org/apache/kafka/metadata/LeaderRecoveryState.java]: Removed the map.

File: clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java (hunk 4 above)
Comment (jsancio) [reply clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:1308]: Updated the test to take into count the version passed.

File: clients/src/main/resources/common/message/AlterPartitionRequest.json (hunk 3 above)
Comment (jsancio) [reply clients/src/main/resources/common/message/AlterPartitionRequest.json]: Fixed both the request and the response.

File: checkstyle/suppressions.xml (hunk 1 above)
Comment (jsancio) [reply checkstyle/suppressions.xml]: Remove this suppression since I fixed the alterPartition method.

File: core/src/test/scala/kafka/zk/TopicPartitionStateZNodeTest.scala (hunk 15 above)
Comment (jsancio) [reply core/src/test/scala/kafka/zk/TopicPartitionStateZNodeTest.scala]: Yes. Added another test.

File: core/src/main/scala/kafka/server/AlterIsrManager.scala (hunk 13 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/server/AlterIsrManager.scala]: Yeah, I think the main thing I didn't like is that `IllegalArgumentException` is such a generic error. So using it to infer a particular case seemed a little slippery.

File: clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java (hunk 2 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java]: I think I'd just get rid of it. Deprecating it sort of sends the wrong signal that it is a public API, when it is definitely not.

File: core/src/test/scala/kafka/zk/TopicPartitionStateZNodeTest.scala [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+    val stat = mock(classOf[Stat])
+    when(stat.getVersion).thenReturn(zkVersion)
+
+    val expected = LeaderIsrAndControllerEpoch(LeaderAndIsr(1, 6, List(1), LeaderRecoveryState.RECOVERING, zkVersion), 10)
+
+    assertEquals(Some(expected), decode(encode(expected), stat))
+  }
+
+  @Test
+  def testEncodeDecodeRecovered(): Unit = {
```
Comment (hachikuji) [inline core/src/test/scala/kafka/zk/TopicPartitionStateZNodeTest.scala:46]: Do we have the case covered when we are decoding an old value without the leader recovery state field?

File: core/src/test/scala/unit/kafka/cluster/PartitionTest.scala [hunk 18]
```diff
@@ -1077,6 +1078,58 @@ class PartitionTest extends AbstractPartitionTest {
 
   }
 
+  @Test
+  def testInvalidAlterPartitionAreNotRetried(): Unit = {
```
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/cluster/PartitionTest.scala]: nit: perhaps `testInvalidAlterPartitionRequestsAreNotRetried`?

File: core/src/test/scala/unit/kafka/cluster/PartitionTest.scala [hunk 19]
```diff
@@ -1077,6 +1078,58 @@ class PartitionTest extends AbstractPartitionTest {
...
+    assertEquals(LogOffsetMetadata.UnknownOffsetMetadata.messageOffset, remoteReplica.logEndOffset)
+    assertEquals(UnifiedLog.UnknownOffset, remoteReplica.logStartOffset)
+
+    partition.updateFollowerFetchState(remoteBrokerId,
+      followerFetchOffsetMetadata = LogOffsetMetadata(10),
+      followerStartOffset = 0L,
+      followerFetchTimeMs = time.milliseconds(),
+      leaderEndOffset = 10L)
+
+    // Check that the isr didn't change and alter update is scheduled
```
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/cluster/PartitionTest.scala:1115]: Could we assert something about the inflight ISR?

File: core/src/test/scala/unit/kafka/cluster/PartitionTest.scala [hunk 20]
```diff
@@ -2015,6 +2067,8 @@ class PartitionTest extends AbstractPartitionTest {
     verify(spyConfigRepository, times(2)).topicConfig(topicPartition.topic())
   }
 
+  // TODO: Add a test that shows that the follower rejects reads until a recovered leader and isr
```
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/cluster/PartitionTest.scala]: Don't forget about TODO

File: metadata/src/main/java/org/apache/kafka/controller/PartitionChangeBuilder.java [hunk 21]
```diff
@@ -104,6 +107,14 @@ public PartitionChangeBuilder setTargetAdding(List<Integer> targetAdding) {
         return this;
     }
 
+    public PartitionChangeBuilder setTargetLeaderRecoveryState(LeaderRecoveryState targetLeaderRecoveryState) {
+        this.targetLeaderRecoveryState = targetLeaderRecoveryState;
+        return this;
+    }
+
+    // TODO: We need to make sure that the LeaderRecoveryState is not lost when the partition transitions from
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/controller/PartitionChangeBuilder.java]: TODO here

File: metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java [hunk 22]
```diff
@@ -760,44 +719,132 @@ BrokersToIsrs brokersToIsrs() {
...
+     */
+    private Errors validateAlterPartitionData(
+        int brokerId,
+        TopicControlInfo topic,
+        int partitionId,
+        PartitionRegistration partition,
+        AlterPartitionRequestData.PartitionData partitionData
+    ) {
+        if (partition == null) {
+            log.info("Rejecting alterPartition request for unknown partition {}-{}.",
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java]: nit: I think we typically capitalize request names in logging

File: metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java [hunk 23]
```diff
@@ -760,44 +719,132 @@ BrokersToIsrs brokersToIsrs() {
...
+                    topic.name, partitionId);
+
+            return UNKNOWN_TOPIC_OR_PARTITION;
+        }
+        if (partitionData.leaderEpoch() != partition.leaderEpoch) {
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:799]: Not part of this patch, but I was wondering if FENCED_LEADER_EPOCH is the right error code when the request epoch is larger than the registration epoch.

File: metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java [hunk 24]
```diff
@@ -760,44 +719,132 @@ BrokersToIsrs brokersToIsrs() {
...
                     }
                 }
-                responseTopicData.partitions().add(new AlterIsrResponseData.PartitionData().
+
+                responseTopicData.partitions().add(new AlterPartitionResponseData.PartitionData().
                     setPartitionIndex(partitionId).
                     setErrorCode(result.code()).
                     setLeaderId(partition.leader).
+                    setIsr(Replicas.toList(partition.isr)).
+                    setLeaderRecoveryState(partition.leaderRecoveryState.value()).
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:766]: Hmm, how do we guarantee that this field is supported?

File: metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java (hunk 23 above)
Comment (jsancio) [reply metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:799]: Yeah. Maybe a INVALID_REQUEST is better. It looks like the legacy controller just processes the request if the leader has a higher epoch then what the controller knows.

File: metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java (hunk 24 above)
Comment (jsancio) [reply metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:766]: It is a tagged field so we can't. I think this is okay for the following reason:
1. If the leader doesn't support this feature it will send version 0 of the AlterPartition request which will set the leader recovery state to RECOVERED
2. The leader only change value if:
  a. a AlterPartition request is sent that increases the ISR
  b. another unclean leader election is performed which sets (keeps) the leader recovery state as RECOVERING

Note that the broker to controller channel uses the ApiVersions response  to determine which version of AlterPartition to send.

File: core/src/test/scala/kafka/zk/TopicPartitionStateZNodeTest.scala (hunk 17 above)
Comment (jsancio) [reply core/src/test/scala/kafka/zk/TopicPartitionStateZNodeTest.scala:46]: We were covering it implicitly. Added another test case that shows the behavior explicitly.

File: core/src/test/scala/unit/kafka/cluster/PartitionTest.scala (hunk 19 above)
Comment (jsancio) [reply core/src/test/scala/unit/kafka/cluster/PartitionTest.scala:1115]: Fixed.

File: core/src/test/scala/unit/kafka/cluster/PartitionTest.scala (hunk 20 above)
Comment (jsancio) [reply core/src/test/scala/unit/kafka/cluster/PartitionTest.scala]: Done. Added this test.

File: metadata/src/main/java/org/apache/kafka/controller/PartitionChangeBuilder.java (hunk 21 above)
Comment (jsancio) [reply metadata/src/main/java/org/apache/kafka/controller/PartitionChangeBuilder.java]: Added two tests:
1. Show that going from online to offline to online preserves the leader recovery state
2. Show that performing an unclean leader election set the leader recovery state to RECOVERING

File: metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java (hunk 22 above)
Comment (jsancio) [reply metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java]: Fixed throughout the file.

File: clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java (hunk 2 above)
Comment (jsancio) [reply clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java]: Done.

File: metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java (hunk 24 above)
Comment (hachikuji) [reply metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:766]: Hmm, it does not seem to be tagged.

Comment (jsancio) [reply metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:766]: Yes. You are correct, The `LeaderRecoveryState` field for `AlterPartitionResponse` is not a tagged field.

The argument for why this is safe is the same as for `KafkaController`. In the success case the `LeaderRecoveryState` in the response is the same as the request. For version 0, this is the default, `RECOVERED`. If the AlterPartition request fails then both controller implementation set the `ErrorCode` field in the response and leave the rest of the fields in their default value.

File: core/src/main/scala/kafka/cluster/Partition.scala (hunk 7 above)
Comment (jsancio) [reply core/src/main/scala/kafka/cluster/Partition.scala:563]: Yes and no. Unclean leader election is not possible when there is only one replica since the ISR set always equals the replica set. That means that if the replica is online then it will be elected as part of regular election and if it is offline the unclean leader election is a noop.

Comment (jsancio) [reply core/src/main/scala/kafka/cluster/Partition.scala:563]: I'll log a message.

File: core/src/main/scala/kafka/cluster/Partition.scala (hunk 8 above)
Comment (jsancio) [reply core/src/main/scala/kafka/cluster/Partition.scala]: I didn't implement this now because only the leader can accept PRODUCE requests and the leader recovers immediately. In this implementation only the follower needs to wait for the leader to recover so it can accept FETCH requests.

**Comment by hachikuji** [comment]
In regard to this [comment](https://github.com/apache/kafka/pull/11733#discussion_r814422959):

> Unclean leader election is not possible when there is only one replica since the ISR set always equals the replica set. That means that if the replica is online then it will be elected as part of regular election and if it is offline the unclean leader election is a noop.

Hmm, suppose we have a partition with replicas=(0) and isr=(0), but 0 is offline. The user chooses to reassign to node 1 and perform an unclean election to make it the leader. That would allow the reassignment to complete, but the partition state would be RECOVERING, right? How do we get out of that state?

**Comment by hachikuji** [comment]
For this [comment](https://github.com/apache/kafka/pull/11733#discussion_r812044247):

> If the version is `0` then this is guaranteed to be the default value `0` so the serialization will succeed. This is true because we only write these values in the response when the operation success. If the operation fails then we skip writing these values and instead just write the error code.

This is a bit subjective, so feel free to disregard, but I do feel like some of the implicit assumptions might be causing some unnecessary obscurity. This is one case where a version check might actually be clearer and prevent the need for the extra comment.
 
A second case is implicitly setting RECOVERED in `PendingPartitionChange`. I had a comment about this [here](https://github.com/apache/kafka/pull/11733/files#r805005385), which might have been missed. This is fine at the moment because the current patch does not do any actual recovery operation, but I think we should reconsider it when we do. Otherwise I do think it's easy to overlook the implication when making other partition state changes.

File: core/src/main/scala/kafka/controller/Election.scala (hunk 9 above)
Comment (jsancio) [reply core/src/main/scala/kafka/controller/Election.scala:49]: Yes. The case that I had in mind is:
1. Leader is elected using unclean leader election. E.g. leader: 1, recoveryState: RECOVERING
2. Leader never sends AlterPartition and goes offline. E.g. leader: -1, recoveryState: RECOVERING
3. Only ISR member (id 1) comes back online. E.g. leader:1, recoveryState: RECOVERING

File: core/src/main/scala/kafka/controller/Election.scala (hunk 10 above)
Comment (jsancio) [reply core/src/main/scala/kafka/controller/Election.scala]: Yes. Fixed. I think I did a search and replace at some point.

File: core/src/main/scala/kafka/controller/KafkaController.scala (hunk 11 above)
Comment (jsancio) [reply core/src/main/scala/kafka/controller/KafkaController.scala]: Fix.

**Comment by jsancio** [comment]
> Hmm, suppose we have a partition with replicas=(0) and isr=(0), but 0 is offline. The user chooses to reassign to node 1 and perform an unclean election to make it the leader. That would allow the reassignment to complete, but the partition state would be RECOVERING, right? How do we get out of that state?

I think that is correct. I think what we can do here is make sure that the leader sends an AlterPartition request when it changes from RECOVERING to RECOVERED. I didn't implement it in this PR since this implementation is a noop for the recovering state. Do you mind if I implement this in a future PR? I filed: https://issues.apache.org/jira/browse/KAFKA-13696

**Comment by jsancio** [comment]
> For this [comment](https://github.com/apache/kafka/pull/11733#discussion_r812044247):
> 
> > If the version is `0` then this is guaranteed to be the default value `0` so the serialization will succeed. This is true because we only write these values in the response when the operation success. If the operation fails then we skip writing these values and instead just write the error code.
> 
> This is a bit subjective, so feel free to disregard, but I do feel like some of the implicit assumptions might be causing some unnecessary obscurity. This is one case where a version check might actually be clearer and prevent the need for the extra comment.
> 
> A second case is implicitly setting RECOVERED in `PendingPartitionChange`. I had a comment about this [here](https://github.com/apache/kafka/pull/11733/files#r805005385), which might have been missed. This is fine at the moment because the current patch does not do any actual recovery operation, but I think we should reconsider it when we do. Otherwise I do think it's easy to overlook the implication when making other partition state changes.

Okay. As you suggested, I marked the LeaderRecoveryState field for the AlterParititon response as ignorable. This has the same behavior we want but implemented in the serialization layer instead of in the controller logic.

**Comment by jsancio** [comment]
@hachikuji I pushed a commit so that both controller implementations only persist the leader recovery state if the cluster supports leader recovery state (IBP is greater than 3.2). I am currently working on the tests.

File: core/src/main/scala/kafka/cluster/Partition.scala (hunk 6 above)
Comment (jsancio) [reply core/src/main/scala/kafka/cluster/Partition.scala:163]: I added `leaderRecoveryState` to the `toString` implementation.

Regarding your other comment, I'll address it in the next PR when I implement https://issues.apache.org/jira/browse/KAFKA-13696

File: core/src/main/scala/kafka/controller/KafkaController.scala (hunk 12 above)
Comment (jsancio) [reply core/src/main/scala/kafka/controller/KafkaController.scala]: Based on this discussion https://github.com/apache/kafka/pull/11733#issuecomment-1051066928, I decided to mark this field as ignorable in the response.

**Review by hachikuji — COMMENTED** [review]
The PR looks good overall. I think there is one problem with the fetch validation. We are expecting that followers will detect the RECOVERED state through a `LeaderAndIsr` request from the controller. However, leaders/followers only accept`LeaderAndIsr` requests if there is an epoch bump, and that does not happen for `AlterPartition` requests. For KRaft, I think it is not a problem.

**Comment by jsancio** [comment]
> The PR looks good overall. I think there is one problem with the fetch validation. We are expecting that followers will detect the RECOVERED state through a `LeaderAndIsr` request from the controller. However, leaders/followers only accept`LeaderAndIsr` requests if there is an epoch bump, and that does not happen for `AlterPartition` requests. For KRaft, I think it is not a problem.

You are correct. I removed the FETCH request validation looking at the leader recovery state and file this issue: https://issues.apache.org/jira/browse/KAFKA-13754

This is okay because at the moment the topic partition leader immediately marks the partition as recovered.

**Review by hachikuji — APPROVED** [review]
Thanks, LGTM
